File: Echoes-of-Entropy/Input/InputManager.cs
using Raylib_cs;

namespace Echoes_of_Entropy.Input;

public class InputManager
{
    private static InputManager? _instance;
    public static InputManager Instance => _instance ??= new InputManager();

    private readonly Dictionary<GameAction, KeyboardKey> _keyBindings = new();

    private InputManager()
    {
        _keyBindings[GameAction.MoveUp] = KeyboardKey.W;
        _keyBindings[GameAction.MoveDown] = KeyboardKey.S;
        _keyBindings[GameAction.MoveLeft] = KeyboardKey.A;
        _keyBindings[GameAction.MoveRight] = KeyboardKey.D;
        _keyBindings[GameAction.Dodge] = KeyboardKey.Space;
    }

    public bool IsActionDown(GameAction action)
    {
        if (_keyBindings.TryGetValue(action, out var key))
        {
            return Raylib.IsKeyDown(key);
        }
        return false;
    }

    public bool IsActionPressed(GameAction action)
    {
        if (_keyBindings.TryGetValue(action, out var key))
        {
            return Raylib.IsKeyPressed(key);
        }
        return false;
    }
}

File: Sources/Input/InputManager.swift
import Raylib

final class InputManager: InputSystem {
    private var keyBindings: [GameAction: KeyboardKey] = [:]

    init() {
        keyBindings[.moveUp] = KeyboardKey.letterW
        keyBindings[.moveDown] = KeyboardKey.letterS
        keyBindings[.moveLeft] = KeyboardKey.letterA
        keyBindings[.moveRight] = KeyboardKey.letterD
        keyBindings[.moveDodge] = KeyboardKey.space
    }

    func isActionDown(_ action: GameAction) -> Bool {
        guard let key = keyBindings[action] else {
            return false
        }
        return Raylib.isKeyDown(key)
    }

    func isKeyPressed(_ action: GameAction) -> Bool {
        guard let key = keyBindings[action] else {
            return false
        }
        return Raylib.isKeyPressed(key)
    }

    func updateState(_ state: inout InputState) {
        for (action, _) in keyBindings {
            state.update(action: action, isActive: isActionDown(action))
        }
    }
}
